tools/rpkg/src/reltoaltrep.cpp

```diff
@@ -0,0 +1,259 @@
+}
+
+void *RelToAltrep::RownamesDataptr(SEXP x, Rboolean writeable) {
+	auto rownames_wrapper = AltrepRownamesWrapper::Get(x);
+	auto row_count = rownames_wrapper->rel->GetQueryResult()->RowCount();
+	if (row_count > (idx_t)NumericLimits<int32_t>::Maximum()) {
+		Rf_error("Integer overflow for row.names attribute");
+	}
+	rownames_wrapper->rowlen_data[1] = -row_count;
+	return rownames_wrapper->rowlen_data;
+}
+
+R_xlen_t RelToAltrep::VectorLength(SEXP x) {
+	return AltrepVectorWrapper::Get(x)->rel->GetQueryResult()->RowCount();
+}
+
+void *RelToAltrep::VectorDataptr(SEXP x, Rboolean writeable) {
+	return AltrepVectorWrapper::Get(x)->Dataptr();
+}
+
+SEXP RelToAltrep::VectorStringElt(SEXP x, R_xlen_t i) {
+	return STRING_ELT(AltrepVectorWrapper::Get(x)->Vector(), i);
+}
+
+static R_altrep_class_t LogicalTypeToAltrepType(LogicalType &type) {
+	switch (type.id()) {
+	case LogicalTypeId::BOOLEAN:
+		return RelToAltrep::logical_class;
+	case LogicalTypeId::UTINYINT:
+	case LogicalTypeId::TINYINT:
+	case LogicalTypeId::SMALLINT:
+	case LogicalTypeId::USMALLINT:
+	case LogicalTypeId::INTEGER:
+		return RelToAltrep::int_class;
+	case LogicalTypeId::UINTEGER:
+	case LogicalTypeId::UBIGINT:
+	case LogicalTypeId::BIGINT:
+	case LogicalTypeId::HUGEINT:
+	case LogicalTypeId::FLOAT:
+	case LogicalTypeId::DOUBLE:
+	case LogicalTypeId::DECIMAL:
+	case LogicalTypeId::TIMESTAMP_SEC:
+	case LogicalTypeId::TIMESTAMP_MS:
+	case LogicalTypeId::TIMESTAMP:
+	case LogicalTypeId::TIMESTAMP_TZ:
+	case LogicalTypeId::TIMESTAMP_NS:
+	case LogicalTypeId::DATE:
+	case LogicalTypeId::TIME:
+	case LogicalTypeId::INTERVAL:
+		return RelToAltrep::real_class;
+	case LogicalTypeId::VARCHAR:
+	case LogicalTypeId::UUID:
+		return RelToAltrep::string_class;
+	default:
+		cpp11::stop("rel_to_altrepUnknown column type for altrep: %s", type.ToString().c_str());
+	}
+}
+
+[[cpp11::register]] SEXP rapi_rel_to_altrep(duckdb::rel_extptr_t rel_p) {
+	D_ASSERT(rel_p && rel_p->rel);
+	auto rel = rel_p->rel;
+	auto ncols = rel->Columns().size();
+
+	cpp11::writable::list data_frame(NEW_LIST(ncols));
+	data_frame.attr(R_ClassSymbol) = RStrings::get().dataframe_str;
+	auto relation_wrapper = make_shared<AltrepRelationWrapper>(rel);
+	RProtector r_protector;
+
+	cpp11::external_pointer<AltrepRownamesWrapper> ptr(new AltrepRownamesWrapper(relation_wrapper));
+	auto row_names_sexp = r_protector.Protect(R_new_altrep(RelToAltrep::rownames_class, ptr, R_NilValue));
+	install_new_attrib(data_frame, R_RowNamesSymbol, row_names_sexp);
+	vector<string> names;
+	for (auto &col : rel->Columns()) {
+		names.push_back(col.Name());
+	}
+
+	SET_NAMES(data_frame, StringsToSexp(names));
+	for (size_t col_idx = 0; col_idx < ncols; col_idx++) {
+		cpp11::external_pointer<AltrepVectorWrapper> ptr(new AltrepVectorWrapper(relation_wrapper, col_idx));
+		auto column_type = rel->Columns()[col_idx].Type();
+		auto vector_sexp = r_protector.Protect(R_new_altrep(LogicalTypeToAltrepType(column_type), ptr, R_NilValue));
```

Comment:
AFAICT protection not needed here?

```suggestion
		auto vector_sexp = R_new_altrep(LogicalTypeToAltrepType(column_type), ptr, R_NilValue);
```

Reply:
we now need it